


## Milestones 1

#### Nama  : Sasi Novia Kirana
#### Batch : CODA-RMT-019

#### Program ini dibuat untuk melakukan proses ETL (Extract, Transform, Load) menggunakan data produk dari website Gramedia


### Persiapan Environment dan Instalasi Package

In [1]:
!pip install -q bs4 selenium pandas


### Import Library , Membuat Web driver , Menentukan url dan melakukan scrolling halaman
    

In [2]:
# data manipulation or dataframe
import time
import pandas as pd

# webscraping
from selenium import webdriver
from bs4 import BeautifulSoup 

# create driver
driver = webdriver.Chrome() 

# link
url = 'https://www.gramedia.com/search?q=novel'

# access the website
driver.get(url) 

# Scroll beberapa kali agar lebih banyak produk muncul 

for i in range(10):
    driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
    


### Mengambil Source HTML dan Membuat Objek BeautifulSoup

In [3]:
# fetch html components
html = driver.page_source 

# parse
soup = BeautifulSoup(html, "html.parser")


In [4]:
print(soup.prettify()[:5000])

<html class="" data-mantine-color-scheme="light" data-sentry-component="Document" data-sentry-element="Html" data-sentry-source-file="_document.tsx" lang="id">
 <head data-sentry-element="Head" data-sentry-source-file="_document.tsx">
  <script async="" src="https://analytics.tiktok.com/i18n/pixel/static/identify_5cff1caf.js" type="text/javascript">
  </script>
  <script async="" data-id="CL0A31RC77U3K90HMAA0" src="https://analytics.tiktok.com/i18n/pixel/static/main.MTc5ZjUxNDU4OQ.js" type="text/javascript">
  </script>
  <script async="" src="https://webtrafficsource.com/track/code.js">
  </script>
  <script async="" src="https://rtg.wewomedia.com/track/code.js">
  </script>
  <script async="" src="https://www.googletagmanager.com/gtag/destination?id=G-FVK74FX33P&amp;cx=c&amp;gtm=4e65t1h1" type="text/javascript">
  </script>
  <script async="" src="https://www.googletagmanager.com/gtag/destination?id=AW-393095129&amp;cx=c&amp;gtm=4e65t1h1" type="text/javascript">
  </script>
  <script

#### List HTML yang diambil dari web gramedia 

In [ ]:
#HTML

judul : 
<h2 class="relative line-clamp-2 text-xs-medium text-neutral-700" title="Novel The Big Four (Empat Besar)" data-testid="productCardTitle">Novel The Big Four (Empat Besar)</h2>
<h2 class="relative line-clamp-2 text-xs-medium text-neutral-700" title="Novel Cantik Itu Luka" data-testid="productCardTitle">Novel Cantik Itu Luka</h2>
<h2 class="relative line-clamp-2 text-xs-medium text-neutral-700" title="Five Nights At Freddy`S Graphic Novel #1: The Silver Eyes" data-testid="productCardTitle">Five Nights At Freddy`S Graphic Novel #1: The Silver Eyes</h2>

Penulis: 
<div class="relative truncate text-2xs-medium text-neutral-500" title="Agatha Christie" data-testid="productCardAuthor">Agatha Christie</div>
<div class="relative truncate text-2xs-medium text-neutral-500" title="Eka Kurniawan" data-testid="productCardAuthor">Eka Kurniawan</div>

Harga:
<div class="text-s-extrabold text-neutral-700" data-testid="productCardFinalPrice">Rp51.750</div>
<div class="text-s-extrabold text-neutral-700" data-testid="productCardFinalPrice">Rp100.000</div>

Diskon: 
<div class="ProductCard_discount__Q8rno inline-block rounded-2xs bg-red-50 px-1 text-2xs-medium text-red-500" data-testid="productCardDiscount">10%</div>
<div class="ProductCard_discount__Q8rno inline-block rounded-2xs bg-red-50 px-1 text-2xs-medium text-red-500" data-testid="productCardDiscount">15%</div>

### Melakukan Looping Data

Untuk mengambil data Judul buku , harga , penulis dan diskon 

In [91]:
# empty list

titles = []
prices = []
authors = []
Discount = []

# Judul

el_titles = soup.find_all('h2', {'class': 'relative line-clamp-2 text-xs-medium text-neutral-700'}) 
for i in el_titles:
    titles.append(i.get_text())

# Harga

el_prices = soup.find_all('div', {'class': 'text-s-extrabold text-neutral-700'}) 
for i in el_prices:
    prices.append(i.get_text())

# Penulis
    
el_authors = soup.find_all('div', {'class': 'relative truncate text-2xs-medium text-neutral-500'})
for i in el_authors:
    authors.append(i.get_text())
    
# Diskon

el_Discount = soup.find_all('div',{'class': 'ProductCard_discount__Q8rno inline-block rounded-2xs bg-red-50 px-1 text-2xs-medium text-red-500'})
for i in el_Discount:
    Discount.append(i.get_text()) 



### Validasi dan Verifikasi Konsistensi Data

In [ ]:
print("judul   :", len(titles))
print("harga   :", len(prices))
print("penulis :", len(authors))
print("diskon  :", len(Discount))

judul   : 221
harga   : 221
penulis : 221
diskon  : 204


### Catatan : 

Setelah dilakukan pengecekan panjang list menggunakan funngsi 'len()' terdapat perbedaan pada jumlah data "diskon" dibandingkan yang lainnya. Penyebabnya tidak semua produk buku dihalaman website gramedia memiliki program diskon. 

### Mengambil seluruh data produk menggunakan find_all()


In [93]:
boxes = soup.find_all('div', {'data-testid': 'productCardContent'}) 

### Melakukan Looping Data 

In [119]:
# empty list

titles = []
prices = []
authors = []
Discount = []

for box in boxes: 
    # Judul 
    try : 
        title = box.find('h2', {'class': 'relative line-clamp-2 text-xs-medium text-neutral-700'}) 
        title = title.get_text(strip=True) if title else None
    except: 
        title = None
    titles.append(title)

    # Harga
    try :
        price = box.find('div', {'class': 'text-s-extrabold text-neutral-700'})
        price = price.get_text(strip=True) if price else None
    except:
        price = None
    prices.append(price) 

    # Penulis
    try:
        author = box.find('div', {'class': 'relative truncate text-2xs-medium text-neutral-500'})
        author = author.get_text(strip=True) if author else None
    except:
        author = None
    authors.append(author)

    # Diskon
    try:
        discount = box.find('div',{'class': 'ProductCard_discount__Q8rno inline-block rounded-2xs bg-red-50 px-1 text-2xs-medium text-red-500'})
        discount = discount.get_text(strip=True) if discount else None
    except:
        discount = None
    Discount.append(discount) 


### Menyimpan Hasil Scraping ke DataFrame

In [120]:
import pandas as pd

df = pd.DataFrame({
    "Judul": titles,
    "Harga": prices,
    "Penulis": authors,
    "Diskon": Discount
})

### Memeriksa hasil scraping

In [121]:
df

,Judul,Harga,Penulis,Diskon
0,Novel The Big Four (Empat Besar),Rp51.750,Agatha Christie,25%
1,Novel Nadira,Rp71.250,Leila S. Chudori,25%
2,Novel Milea Suara Dari Dilan,Rp80.100,Pidi Baiq,10%
3,Novel Critical Eleven,Rp74.250,Ika Natassa,25%
4,Novel Dilan 2:Dia Adalah Dilanku Tahun 1991,Rp89.100,Pidi Baiq,10%
...,...,...,...,...
216,Disney Pixar Elemental: The Junior Novel,Rp68.400,Igloo Books,10%
217,Hook Your Readers: Langkah Pertama Menulis Novel,Rp94.500,Arai Hisayuki,10%
218,Seni Perang Sun Tzu: Sebuah Novel Grafis,Rp108.000,Pete Katz,NaN
219,Isu Keliyanan dan Resistensi dalam Ruang Kolon...,Rp81.000,Wawat Rahwati,10%


In [122]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 221 entries, 0 to 220
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Judul    221 non-null    str  
 1   Harga    221 non-null    str  
 2   Penulis  221 non-null    str  
 3   Diskon   204 non-null    str  
dtypes: str(4)
memory usage: 7.0 KB


##### Dari hasil pengecekan didapat : 

    1. Dataset terdiri dari 221 baris dan 4 kolom.
    2. Kolom Judul, Harga, dan Penulis tidak memiliki missing value karena seluruh 221 data terisi.
    3. Kolom Diskon memiliki 204 data terisi dan 17 data kosong (missing value).
    4. Seluruh kolom masih bertipe data string (str).
    5. Kolom Harga dan Diskon seharusnya bertipe numerik sehingga perlu dilakukan proses transformasi dengan menghapus karakter non-angka dan mengubah tipe data menjadi integer.

### Pembersihan dan Transformasi Kolom Diskon

Pada tahap ini dilakukan pembersihan data pada kolom `Diskon`. Simbol `%` dihapus, nilai `NaN` diganti menjadi `0`, kemudian seluruh data dikonversi menjadi tipe numerik (`integer`) agar dapat digunakan untuk analisis dan penyimpanan ke database.

In [123]:
df['Diskon'] = (
    df['Diskon']
    .astype(str)
    .str.replace('%', '', regex=False)
    .replace('None', '0')
)

df['Diskon'] = pd.to_numeric(df['Diskon'], errors='coerce').fillna(0).astype(int)

### Menampilkan dataset setelah perubahan pada bagian diskon

In [124]:
df

,Judul,Harga,Penulis,Diskon
0,Novel The Big Four (Empat Besar),Rp51.750,Agatha Christie,25
1,Novel Nadira,Rp71.250,Leila S. Chudori,25
2,Novel Milea Suara Dari Dilan,Rp80.100,Pidi Baiq,10
3,Novel Critical Eleven,Rp74.250,Ika Natassa,25
4,Novel Dilan 2:Dia Adalah Dilanku Tahun 1991,Rp89.100,Pidi Baiq,10
...,...,...,...,...
216,Disney Pixar Elemental: The Junior Novel,Rp68.400,Igloo Books,10
217,Hook Your Readers: Langkah Pertama Menulis Novel,Rp94.500,Arai Hisayuki,10
218,Seni Perang Sun Tzu: Sebuah Novel Grafis,Rp108.000,Pete Katz,0
219,Isu Keliyanan dan Resistensi dalam Ruang Kolon...,Rp81.000,Wawat Rahwati,10


### Pembersihan dan Transformasi Kolom Harga

Pada tahap ini dilakukan pembersihan data pada kolom `Harga`. Karakter non-numerik seperti simbol `Rp` dan tanda titik (`.`) sebagai pemisah ribuan dihapus. Setelah itu, data dikonversi menjadi tipe numerik agar dapat digunakan untuk analisis dan penyimpanan ke database PostgreSQL.

In [ ]:
df['Harga'] = (
    df['Harga']
    .str.replace('Rp', '', regex=False)
    .str.replace('.', '', regex=False)  
    .str.strip()
)

df['Harga'] = pd.to_numeric(df['Harga'], errors='coerce')

### Menampilkan dataset setelah perubahan pada bagian harga 

In [129]:
df

,Judul,Harga,Penulis,Diskon
0,Novel The Big Four (Empat Besar),51750,Agatha Christie,25
1,Novel Nadira,71250,Leila S. Chudori,25
2,Novel Milea Suara Dari Dilan,80100,Pidi Baiq,10
3,Novel Critical Eleven,74250,Ika Natassa,25
4,Novel Dilan 2:Dia Adalah Dilanku Tahun 1991,89100,Pidi Baiq,10
...,...,...,...,...
216,Disney Pixar Elemental: The Junior Novel,68400,Igloo Books,10
217,Hook Your Readers: Langkah Pertama Menulis Novel,94500,Arai Hisayuki,10
218,Seni Perang Sun Tzu: Sebuah Novel Grafis,108000,Pete Katz,0
219,Isu Keliyanan dan Resistensi dalam Ruang Kolon...,81000,Wawat Rahwati,10


### Pengecekan tipe data 

In [132]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 221 entries, 0 to 220
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Judul    221 non-null    str  
 1   Harga    221 non-null    int64
 2   Penulis  221 non-null    str  
 3   Diskon   221 non-null    int64
dtypes: int64(2), str(2)
memory usage: 7.0 KB


##### Hasil pengecekan menunjukkan bahwa kolom `Harga` dan `Diskon` telah berhasil dikonversi menjadi tipe data numerik (`int64`). Selain itu, seluruh kolom memiliki jumlah data yang lengkap tanpa missing value sehingga dataset siap disimpan ke file CSV dan dimuat ke PostgreSQL.

### Menyimpan Data Hasil Transformasi ke CSV

In [130]:
df.to_csv("novel_gramedia.csv", index=False)

### Menutup Web Driver 

In [131]:
driver.quit()